scripts/igmp_query_injector.py

```diff
@@ -1,232 +1,95 @@
-
-    def mark_injected(self):
-        self._injected = True
-
-    @property
-    def injected(self):
-        return self._injected
-
-
-class IGMPQueryGenerator(object):
-    """IGMP Query generator.
-    Generate IGMP Query packet with/without vlan
-    """
-    def __init__(self, dst_mac, vlanid, max_resp_time):
-        """
-        :param dst_mac: Destination mac address of this IGMP query packet
-        :param vlanid: Vlan ID of this packet. `-1` means no vlan
-        :param max_resp_time: Max response time of IGMP query. Unit is 100ms
-        """
-        self.src_mac = '00:00:00:00:00:00'
-        self.dst_mac = dst_mac
-        self.vlanid = vlanid
-        self.max_resp_time = max_resp_time
-
-    def mac2binarry(self, mac):
-        return binascii.unhexlify(mac.replace(':', ''))
-
-    def create_ether_layer(self):
-        if self.vlanid == -1:
-            type_field = 0x0800
-        else:
-            type_field = 0x8100
-        return struct.pack('!6s6sH', self.mac2binarry(self.dst_mac), self.mac2binarry(self.src_mac), type_field)
-
-    def create_vlan_layer(self):
-        if self.vlanid == -1:
-            return struct.pack('')
-        else:
-            return struct.pack('!HH', 0x2000 | self.vlanid, 0x0800)
-
-    def create_ip_layer(self):
-        """
-        IP Type: IPv4 (0x0800)
-        Version: 4
-        Total length: 32
-        TTL: 1
-        Source IP: 0.0.0.0
-        Destination IP: 224.0.0.1 (e0:00:00:01)
-        """
-        hex_str = "4600002000010000010244d600000000e000000194040000"
-        s = binascii.unhexlify(hex_str)
-        return struct.pack('!%ds' % len(s), s)
-
-    def create_igmp_layer(self):
-        """
-        IGMP Version: 2
-        IGMP Type: IGMP Query (0x11)
-        Max Resp Time: self.max_resp_time
-        Multicast Address: 0.0.0.0
-        """
-        msg_type = 0x11
-        # MAX_RESP_TIME filed unit is 100ms
-        max_resp_time = self.max_resp_time / 100
-        group_address = socket.inet_aton('0.0.0.0')
-        chksum = self._calc_igmp_checksum(struct.pack('!BBH4s', msg_type, max_resp_time, 0x0000, group_address))
-        return struct.pack('!BBH4s', msg_type, max_resp_time, chksum, group_address)
-
-    def _calc_igmp_checksum(self, packet):
-        def carry_around_add(a, b):
-            c = a + b
-            return (c & 0xffff) + (c >> 16)
-
-        s = 0
-        msg = struct.unpack('%dB' % len(packet), packet)
-        for i in range(0, len(msg), 2):
-            w = (msg[i] << 8) + msg[i+1]
-            s = carry_around_add(s, w)
-        return ~s & 0xffff
-
-    def generate(self):
-        ether_layer = self.create_ether_layer()
-        vlan_layer = self.create_vlan_layer()
-        ip_layer = self.create_ip_layer()
-        igmp_layer = self.create_igmp_layer()
-        packet = ether_layer + vlan_layer + ip_layer + igmp_layer
-        return packet
+        domid, vifid = [int(x) for x in self.vif_name.split('vif')[1].split('.')]
```

Comment:
Please define helper functions to calculate these values.

Reply:
Accept